### DISCUSSION TITLE
Managing large resources?

### DISCUSSION THREAD

**Discussion by bobh66** [nolar/kopf#1047] [Q&A] [answered]
I'm trying to manage some resources that can be very large - - more than the 262144 bytes allowed in annotations for the last-changed-config, so I'm getting 422 Unprocessable Entity errors when kopf tries to update the resource status.

```
'causes': [{'reason': 'FieldValueTooLong', 'message': 'Too long: must have at most 262144 bytes', 'field': 'metadata.annotations'}, {'reason': 'FieldValueTooLong', 'message': 'Too long: must have at most 262144 bytes', 'field': 'metadata.annotations'}, {'reason': 'FieldValueTooLong', 'message': 'Too long: must have at most 262144 bytes', 'field': 'metadata.annotations'}]}, 'code': 422})
```

Is there a solution/workaround other than server-side-apply, which seems like a lot of work to implement?

Thanks

**Comment by nolar** [comment, marked as answer]
You can implement your own state & progress key-value storages (see the docs), e.g. in sqlite/postgres/mysql, or in a shared filesystem with json in files. 

Or you can configure the pre-implemented storages in the status field (also documented). 

The annotation storages are made a default to serve builtin resources (like pods) out of the box, since the builtins prohibit writing arbitrary data to their status. The same can be expected for crds unless their schema allows such storage.

**Comment by bobh66** [comment]
Thanks @nolar - that worked.

For reference this is what I used:

```
@kopf.on.startup()
async def configure(settings: kopf.OperatorSettings, memo: kopf.Memo, logger, **_):
    settings.watching.client_timeout = 610
    settings.watching.connect_timeout = 60
    settings.watching.server_timeout = 600

    settings.persistence.progress_storage = kopf.StatusProgressStorage()
    settings.persistence.diffbase_storage = kopf.StatusDiffBaseStorage()
```
I have `x-kubernetes-preserve-unknown-fields: true` on the status field in my CRDs
